### ISSUE TITLE
API to tell OPA to download a new bundle

### ISSUE THREAD

**Issue by spellr** [open-policy-agent/opa#1336]
Right now, when changing a bundle, we need to wait for all OPAs to download the new bundle periodically, and update their policies. This can raise security issues, and also make neighboring OPAs to have conflicting states/decisions.

A possible work-around can be shortening the download frequency, but that can create constant load on the bundle server. 

It would be great to have an API endpoint in OPA to tell it, "update now".
This way, when a new policy is added to the bundle, we can tell all OPAs to download the new bundle.

**Comment by srenatus** [comment]
I suppose there's alternatives serving the same goal? I.e., some kind of _subscription mechanism_ for bundle downloading? (like using long polling, websockets, sse, ... whatever's appropriate)

**Comment by tsandall** [comment]
Like @srenatus mentions, this is somewhat related to #1055.

One option would be to extend the OPA API to include a bundle resource.

```
POST /v1/bundles/<name>/sync    # trigger bundle synchronization
```

I could imagine it being useful to have an API that returns the bundle status (activated, errors, timestamps, etc.)

```
GET /v1/bundles/<name>          # return specific bundle status
GET /v1/bundles                 # return list of bundle status
```

@timothyhinrichs WDYT?

**Comment by Devlin556** [comment]
+1, very important feature.It really needed.

**Comment by UWellFail** [comment]
+1

**Comment by kolyasya** [comment]
+1 please add asap

**Comment by tsandall** [comment]
I'm hesitant to go ahead and implement this when it overlaps heavily with #1055. Curious what @timothyhinrichs and @koponen-styra have to say.

**Comment by koponen-styra** [comment]
Accelerating the download with an explicit signal from outside sounds good to me! I have no strong options about the involved API change(s), though.

**Comment by timothyhinrichs** [comment]
I'd say #1055 is the right approach for production.  The bundle API is
partly designed to handle the case where OPA is running as part of a k8s
deployment or auto-scaling-group, so OPA instances spin up and down
automatically.  To work with that same model, running a side-car next to
OPA that finds out about policy updates and tells the OPA API to pull the
latest bundle would probably open an outbound, long-lived connection;
whenever it receives a response, it would tell OPA to start the bundle
download.  That's basically what #1055 is doing but with delta encodings to
minimize download time.

That said, I could see an API for "pull bundle now" being useful for
debugging.  But for that case I've just set the polling interval to 10s.

**Comment by gujun4990** [comment]
Currently, @ashutosh-narkar have realized the long polling(https://github.com/open-policy-agent/opa/pull/3418).

**Comment by multigrid101** [comment]
Hello all, was this ever implemented? I can can see config variables for this in the docs (e.g. `bundles[_].trigger = manual`) but no info on how to actually execute the manual trigger.

**Comment by anderseknert** [comment]
@multigrid101 no, the issue is still open :) The manual trigger you're refering to is a feature of the Go SDK, which I believe was contributed for the purpose of running OPA in AWS Lambdas, where periodic polling is undesirable. It's not currently exposed via the HTTP API.

**Comment by renatosc** [comment]
I second this. The "Delta Bundle" [#1055](https://github.com/open-policy-agent/opa/issues/1055) doesn't support policy change, only data update. Having a API that forces the OPA to download the bundle right away would be great.

**Comment by anderseknert** [comment]
@renatosc long polling (see my previous comment) works for any type of bundles, if that's an option for you.

**Comment by renatosc** [comment]
@anderseknert , yes we have short and long pooling but on either options you are overloading the server (via multiple requests or by keeping a connection open). If you already have a system working on a subscription based model that receive events, having to leave another connection open is a waste of resource.

I don't mind forking and submitting a PR for the change.

**Comment by charlieegan3** [comment]
> on either options you are overloading the server

Out of interest, is your server setting Etag headers in responses when downloading bundles?

**Comment by renatosc** [comment]
Hi @charlieegan3, we do not have implemented it yet. When I said "we have" it was referring to that I know we have the options of using short/long pooling with OPA right now.  Our system will have a SignalR connection to receive events such as OPA needs to update its bundle. So having OPA using short/long polling would be unnecessary if we had a way to trigger the bundle sync/refresh as suggested by this issue.

**Comment by charlieegan3** [comment]
My concern in your scenario @renatosc is that by triggering reloading of bundles via an API would create a [Thundering Herd Problem](https://en.wikipedia.org/wiki/Thundering_herd_problem) when many OPAs update at the same time. This might be even more problematic if you're concerned about your bundle server load.

While I can see some value in adding this endpoint for testing purposes, I remain unconvinced that it's a good pattern for most production use cases or achieving consistency in loaded bundle state of many OPA instances. Oftentimes, new OPA instances can start at any moment - either as sidecars or from an horizontal scaling event.

If we do add this feature, I think it'd be good to document that relying on it would be a step away from a [level based](https://stackoverflow.com/questions/31041766/what-does-edge-based-and-level-based-mean) system which is often a desirable property.

**Comment by ashutosh-narkar** [comment]
@renatosc FWIW there are currently two ways to control how a bundle is downloaded from the remote server. The default is `periodic` wherein OPA periodically downloads a bundle. And there is the `manual` mode where you can manually trigger the bundle download. This feature is exposed via OPA's GO API atm.
